src/toast/ops/groundfilter.py

```diff
@@ -0,0 +1,373 @@
+    def _check_det_flag_mask(self, proposal):
+        check = proposal["value"]
+        if check < 0:
+            raise traitlets.TraitError("Flag mask should be a positive integer")
+        return check
+
+    @traitlets.validate("shared_flag_mask")
+    def _check_shared_flag_mask(self, proposal):
+        check = proposal["value"]
+        if check < 0:
+            raise traitlets.TraitError("Flag mask should be a positive integer")
+        return check
+
+    @traitlets.validate("trend_order")
+    def _check_trend_order(self, proposal):
+        check = proposal["value"]
+        if check < 0:
+            raise traitlets.TraitError("Trend order should be a non-negative integer")
+        return check
+
+    @traitlets.validate("filter_order")
+    def _check_filter_order(self, proposal):
+        check = proposal["value"]
+        if check < 0:
+            raise traitlets.TraitError("Filtere order should be a non-negative integer")
+        return check
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    @function_timer
+    def build_templates(self, obs):
+        """Construct the local ground template hierarchy"""
+
+        views = obs.view[self.view]
+
+        # Construct trend templates.  Full domain for x is [-1, 1]
+
+        my_offset = obs.local_index_offset
+        my_nsamp = obs.n_local_samples
+        nsamp_tot = obs.n_all_samples
+        x = np.arange(my_offset, my_offset + my_nsamp) / nsamp_tot * 2 - 1
+
+        # Do not include the offset in the trend.  It will be part of
+        # of the ground template
+        legendre_trend = np.zeros([self.trend_order, x.size])
+        legendre(x, legendre_trend, 1, self.trend_order + 1)
+
+        try:
+            azmin = obs["scan_min_az"].to_value(u.radian)
+            azmax = obs["scan_max_az"].to_value(u.radian)
+            if self.azimuth is not None:
+                az = obs.shared[self.azimuth]
+            else:
+                quats = obs.shared[self.boresight_azel]
+                theta, phi = qa.to_position(quats)
+                az = 2 * np.pi - phi
+        except Exception as e:
+            raise RuntimeError(
+                "Failed to get boresight azimuth from TOD.  Perhaps it is "
+                'not ground TOD? "{}"'.format(e)
+            )
+
+        # The azimuth vector is assumed to be arranged so that the
+        # azimuth increases monotonously even across the zero meridian.
+
+        phase = (az - azmin) / (azmax - azmin) * 2 - 1
+        nfilter = self.filter_order + 1
+        legendre_templates = np.zeros([nfilter, phase.size])
+        legendre(phase, legendre_templates, 0, nfilter)
+        if not self.split_template:
+            legendre_filter = legendre_templates
+        else:
+            # Create separate templates for alternating scans
+            common_flags = obs.shared[self.shared_flags].data
+            legendre_filter = []
+            # The flag masks are hard-coded in sim_ground.py
```

Comment:
In the future, there will be many ways of populating Observations with both real and simulated data.  This is fine for now, but we will need to make the left and right-going flags / intervals be a configurable parameter eventually.  This goes beyond a simple port of toast2 code though, so fine to leave for now.

Reply:
I agree.